src/adrs/0129349218/README.md

```diff
@@ -0,0 +1,540 @@
+a decent level of craftsmanship compared to starting over all the time; applying
+lessons learned and taking into consideration moderns way to build software.
+
+### Out of Scope
+
+These are few things that are out of scope for this specification, but worth
+mentioning, this is not an exhaustive list, and we would like to hear your
+opinion about it.
+
+- Performance Considerations
+- Handling Sensitive Data
+
+### Decision Drivers:
+
+#### Data-Oriented
+
+Starting the specification as if an error is a data structure that could be
+serialized and deserialized is a good way to start.
+
+System exceptions should be treated it the same. Although, wrapping system
+exceptions seems to be a lot of work, barely system exceptions are intended to
+be bubbled up to the end-user, but instead, they are meant to be handled by the
+application to provide a better user experience and also to provide more context
+to the error.
+
+Errors are just another Message in the system, agnostic of programming language
+or framework details.
+
+#### Identity
+
+In an era of distributed computing and software meant to be interactive across
+multiple actors such as Customers, Customer Services, Developers, SREs or even
+Cloud Providers.
+
+Somewhere in the chain communicating downstream closer to where the error
+happened will be inevitable and the communication among all those actors with a
+shared context having an identity is crucial since they could all point to the
+same entity that caused the issue or at least have a common understanding of
+what happened.
+
+This is also important for debugging purposes. Having a common identity for the
+error will allow you to search for it across multiple systems and have a better
+understanding of the impact of the error.
+
+Having an identity for the error will allow you to search for it across multiple
+systems and have a better understanding of the impact of the error.
+As well as opens the possibility to show the same error to the end-user and
+end-user sharing the identity of the error with the support team.
+
+The identity should avoid distributed systems coordination, as much as
+possible. The identity should be composed of multiple dimensions to allow you
+to have more consistency across systems without losing specificity of a given
+context.
+
+- `domain`: The logical grouping to which the "code" belongs. The error domain
+  is typically the registered service name of the tool or product that
+  generates the error. Example: "com.myapp.iam". If the error is
+  generated by some common infrastructure, the error domain must be a
+  globally unique value that identifies the infrastructure.
+- `code`: The code of the error. This is a constant value that identifies
+  the proximate cause of the error. Error codes are unique within a particular
+  domain of errors.
+- `id`: The ID of an instance of the error. It could be a UUID, Nanoid, or any
+  other unique identifier that could be generated without coordination.
```

Comment:
Thinking specifically of systems where there can be multiple instances of the same service (and in some systems, there can be temporary drift between instances during rollover)...

Would it make sense to have an optional identifier for the service itself (more specific than the domain)?

For example, if we have two instances of `com.myapp.iam`, and during rollover (or blue-green), we encounter an error, I think having that additional detail of a unique identifier per instance will be useful. The identifier itself could be anything (ARN for AWS, commit/release SHA for K8s, or anything else that fits the implementation).

---

Separate question:
Is it intended for `domain` to include the environment? Or is the assumption that environment is out of scope here?

Reply:
There are multiple ways to deal with the situation. Ideally, the "internal metadata" would allow you to tag your error with any data you wish to have.

The `domain` is intended to be as static as possible and define the Business Domain or Infrastructure Domain.

I wouldn't include the environment, but it is up to you; its meaning is based on the **system**.

It Depends 🤷🏻 

The dimensions of the payload, which is my primary focus, kind of cover all of the above. I intend to standardize some of the values in another ADR unless it is something that we can for sure tell will be the same across different businesses.

Reply:
That makes sense. I think that in my own systems, I likely would include some sort of release identifier/version number in some standard form that is present across all errors (but not in domain, since that should be static, like you said).

After thinking more about `environment` I think it's a totally separate issue, as I don't think that logs between environments shouldn't be mixed, so no need to include it.